forge/db/controllers/User.js

```diff
@@ -91,6 +91,67 @@ module.exports = {
         throw new Error('Invalid link')
     },
 
+    generatePendingEmailChangeToken: async function (app, user, newEmailAddress) {
+        const TOKEN_EXPIRY = 1000 * 60 * 60 * 24 * 2 // 48 Hours
+        const expiresAt = Math.floor((Date.now() + TOKEN_EXPIRY) / 1000) // 48 hours
+        const signingHash = sha256(user.password)
+        return jwt.sign({ sub: user.email, id: user.hashid, change: newEmailAddress, aud: 'update-user-email', exp: expiresAt }, signingHash)
+    },
+
+    sendPendingEmailChangeEmail: async function (app, user, newEmailAddress) {
+        const pendingEmailChangeToken = await app.db.controllers.User.generatePendingEmailChangeToken(user, newEmailAddress)
+        await app.postoffice.send(
+            user,
+            'PendingEmailChange',
+            {
+                oldEmail: user.email,
+                newEmail: newEmailAddress,
+                confirmEmailLink: `${app.config.base_url}/account/email_change/${pendingEmailChangeToken}`
+            }
+        )
+    },
+
+    applyPendingEmailChange: async function (app, user, token) {
```

Comment:
You validate the new email is available before generating and sending the link.

We should revalidate that here as some time will have passed and the email may no longer be available.

Reply:
> We should revalidate that here as some time will have passed and the email may no longer be available.

Agreed & is already the case (implicitly) by virtue of the `isEmal` validator on the table column.

I can be explicit (or add a comment) if you wish Nick?